### PR TITLE
caddy: Cross-compilation fix, various enhancements

### LINKED ISSUE #310067: caddy fails to build on platform riscv64
### Describe the bug

Caddy fails to build for riscv64

```
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > Running phase: buildPhase
       > Building subPackage ./cmd/caddy
       > buildPhase completed in 41 seconds
       > Running phase: installPhase
       > /nix/store/75lnjrxbkj4s6mrd77lqnck64aik88sx-stdenv-linux/setup: line 119: /nix/store/gbvs2sikhr4rn6s5baxwy43y27vbq30s-caddy-riscv64-unknown-linux-gnu-2.7.6/bin/caddy: cannot execute binary file: Exec format error
       > /nix/store/75lnjrxbkj4s6mrd77lqnck64aik88sx-stdenv-linux/setup: line 131: pop_var_context: head of shell_variables not a function context
```

### Steps To Reproduce
1. ...
2. ...
3. ...

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/310067

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by OPNA2608** [comment]
@ofborg build pkgsCross.riscv64.caddy

File: pkgs/by-name/ca/caddy/package.nix [hunk 1]
```diff
@@ -40,8 +41,13 @@ buildGoModule {
   postInstall = ''
     install -Dm644 ${dist}/init/caddy.service ${dist}/init/caddy-api.service -t $out/lib/systemd/system
 
-    substituteInPlace $out/lib/systemd/system/caddy.service --replace "/usr/bin/caddy" "$out/bin/caddy"
-    substituteInPlace $out/lib/systemd/system/caddy-api.service --replace "/usr/bin/caddy" "$out/bin/caddy"
+    substituteInPlace $out/lib/systemd/system/caddy.service \
+      --replace-fail "/usr/bin/caddy" "$out/bin/caddy"
+    substituteInPlace $out/lib/systemd/system/caddy-api.service \
+      --replace-fail "/usr/bin/caddy" "$out/bin/caddy"
+  '' + lib.optionalString (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
```
Comment (OPNA2608) [inline pkgs/by-name/ca/caddy/package.nix:48]: This is currently inverted ("if I cannot execute binaries of the platform I'm building for: execute the binary I just built").
```suggestion
  '' + lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]

**Comment by techknowlogick** [comment]
Thank you for this PR, could you split it up into two commits? One to move caddy and xcaddy into `by-name`, and another for the riscv64 fix (perhaps even have the commit be "disable manpage and shell completion on cross-compile", or something similarly descriptive)?

**Comment by onny** [comment]
Okay done :)

**Review by OPNA2608 — APPROVED** [review]
Looks okay to me, riscv64 cross passes on ofborg.

**Comment by onny** [comment]
Result of `nixpkgs-review pr 310561` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 test built:</summary>
  <ul>
    <li>nixosTests.caddy</li>
  </ul>
</details>

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**